docs/concepts/custom-ui-headless-api.mdx

```diff
@@ -0,0 +1,662 @@
+
+<CodeBlock language="go">{createLoginApi}</CodeBlock>
+```
+
+</TabItem>
+<TabItem value="typescript" label="TypeScript">
+
+```mdx-code-block
+import createTsLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/create-login-native.ts'
+
+<CodeBlock language="ts">{createTsLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+<TabItem value="react" label="React">
+
+```mdx-code-block
+import createReactLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/create-login-react.tsx'
+
+<CodeBlock language="ts">{createReactLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+
+</Tabs>
+````
+
+#### Two-Factor authentication
+
+Two-Factor authentication (2FA) is a second step to confirming the user's credentials with another credential, such as `TOTP` or
+`TouchID`. When creating a login flow, you can specify the `?aal=<level>` query parameter to specify the level of authentication.
+
+In Ory the levels are:
+
+- aal1: Password or OIDC
+- aal2: aal1 + TOTP or WebAuthn or Lookup Secrets
+
+In the case of initializing a second factor, the user must already have a valid session cookie. The
+[`/sessions/whoami`](../reference/api#tag/frontend/operation/toSession) endpoint will return with an error with an id
+`session_aal2_required` if the user is required to complete a second factor. More information on UI error messages can be found on
+the [Customize user interface](../kratos/concepts/ui-user-interface) page.
+
+For more information on Multi-factor authentication take a look at the [Multi-factor authentication](../kratos/mfa/overview).
+
+### Get Login flow
+
+````mdx-code-block
+<Tabs groupId="sdk-examples">
+
+<TabItem value="go" label="Go" default>
+
+```mdx-code-block
+import getLoginApi from '!!raw-loader!../../code-examples/sdk/go/selfservice/get-login.go'
+
+<CodeBlock language="go">{getLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+<TabItem value="typescript" label="TypeScript">
+
+```mdx-code-block
+import getTsLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/get-login-native.ts'
+
+<CodeBlock language="ts">{getTsLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+<TabItem value="react" label="React">
+
+```mdx-code-block
+import getReactLoginApi from '!!raw-loader!../../code-examples/sdk/typescript/src/selfservice/login/get-login-react.tsx'
+
+<CodeBlock language="ts">{getReactLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+</Tabs>
+````
+
+### Submit Login flow
+
+We can submit the login flow with user data (e.g. username and password) using the flow ID. Depending on the type of application
+(e.g. browser or native) the CSRF tokens are required in the request body and cookie.
+
+In some cases, such as submitting WebAuthn passwordless, the response code could be `422` which indicates that the browser needs
+to be redirected with a newly created flow. In Single Page Applications (SPAs) instead of redirecting, the response can be decoded
+and the new flow ID extracted.
```

Comment:
This concept of JSON payload vs redirect is true for basically all API calls and should be explained way earlier, not in this subsubsection.